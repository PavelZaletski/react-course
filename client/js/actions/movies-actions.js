import fetch from 'cross-fetch';

export const FETCH_MOVIES_FULFILLED = 'FETCH_MOVIES_FULFILLED';
export const FETCH_MOVIES_REQUEST = 'FETCH_MOVIES_REQUEST';
export const SORT_MOVIES = 'SORT_MOVIES';
const url = 'http://react-cdp-api.herokuapp.com/movies'

export const fetchMovies = (params) => (dispatch) => {
	dispatch(requestMovies())
	let urlWithParams = url + '?limit=12';

	for (let key in params) {
		const value = params[key];
		urlWithParams += `&${key}=${value}`;
	}

	return fetch(urlWithParams, {
			method: 'get'
		})
		.then(res => res.json())
		.then(({data}) =>
			dispatch(moviesFetched(data))
		);
}

export const moviesFetched = (payload) => (
	{ type: FETCH_MOVIES_FULFILLED, payload }
);

export const requestMovies = () => (
	{ type: FETCH_MOVIES_REQUEST }
);

export const sortMovies = (payload) => (
	{ type: SORT_MOVIES, payload }
);
