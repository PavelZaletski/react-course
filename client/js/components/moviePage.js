import React from 'react';
import { TopContainer } from './topContainer';

export class MoviePage extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        const { movie } = this.props;

        return (
            movie ?
                <TopContainer>
                    <div className="movie-page__info">
                        <div className="movie-page__img-wrapper">
                            <img src={movie.poster_path} className="movie-page__img"/>
                        </div>
                        <div className="movie-page__description">
                            <h2 className="movie-page__title">{movie.title}</h2>
                            <div className="movie-page__tagline"> {movie.tagline}</div>
                            <span className="movie-page__year">{new Date(movie.release_date).getFullYear()}</span>
                            { movie.runtime && <span className="movie-page__runtime">{`${movie.runtime} min`}</span> }
                            <div className="movie-page__overview"> {movie.overview}</div>
                        </div>
                    </div>
                </TopContainer>
            :
            null
        );
    }
}