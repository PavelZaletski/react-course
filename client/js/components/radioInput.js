import React from 'react';
import PropTypes from 'prop-types';

export class RadioInput extends React.Component {
    static propTypes = {
        name: PropTypes.string.isRequired,
        value: PropTypes.string.isRequired,
        onChange: PropTypes.func.isRequired
    };

    onChange = (e) => {
        this.props.onChange(e.target.value);
    }

    render() {
        const className = `${this.props.checked ? 'checked' : ''}`;

        return (
            <label className={className}>
                <input name={this.props.name} type="radio" onChange={this.onChange} value={this.props.value} />
                {this.props.text || this.props.value}
            </label>
        );
    }
}