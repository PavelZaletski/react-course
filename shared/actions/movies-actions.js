import { call, put, all, takeLatest } from 'redux-saga/effects';

export const FETCH_MOVIES_FULFILLED = 'FETCH_MOVIES_FULFILLED';
export const FETCH_MOVIES_REQUEST = 'FETCH_MOVIES_REQUEST';
export const FETCH_MOVIE_REQUEST = 'FETCH_MOVIE_REQUEST';
export const SORT_MOVIES = 'SORT_MOVIES';
export const FETCH_MOVIE_FULFILLED = 'FETCH_MOVIE_FULFILLED';
export const FETCH_RELATED_MOVIES_FULFILLED = 'FETCH_THE_SAME_GENRE_MOVIES_FULFILLED';
export const FETCH_REJECTED = 'FETCH_REJECTED';
export const SEARCH_BY = 'SEARCH_BY';
export const url = 'http://react-cdp-api.herokuapp.com/movies';

let params;
let id;

export const moviesFetched = payload => (
  { type: FETCH_MOVIES_FULFILLED, payload }
);

export const movieFetched = payload => (
  { type: FETCH_MOVIE_FULFILLED, payload }
);

export function* fetchMoviesAsync() {
  let urlWithParams = `${url}?limit=12`;

  for (const key in params) {
    const value = params[key];
    urlWithParams += `&${key}=${value}`;
  }

  const response = yield call(fetch, urlWithParams);
  const movies = yield response.json();

  yield put(moviesFetched(movies.data));
}

export function* watchFetchMovies() {
  yield takeLatest(FETCH_MOVIES_REQUEST, fetchMoviesAsync);
}

export function* fetchMovieByIdAsync() {
  const response = yield call(fetch, `${url}/${id}`);
  const movie = yield response.json();

  const newUrl = `${url}?limit=12&searchBy=genres&search=${movie.genres[0]}`;

  const response2 = yield call(fetch, newUrl);
  const relatedMovies = yield response2.json();

  yield put(movieFetched({ movie, relatedMovies: relatedMovies.data }));
}

export function* watchFetchMovieById() {
  yield takeLatest(FETCH_MOVIE_REQUEST, fetchMovieByIdAsync);
}

export function* moviesSaga() {
  yield all([
    watchFetchMovies(),
    watchFetchMovieById(),
  ]);
}

export const fetchMovies = (_params) => {
  params = _params;
  return { type: FETCH_MOVIES_REQUEST };
};

export const fetchMovieById = (_id) => {
  id = _id;
  return { type: FETCH_MOVIE_REQUEST };
};

export const sortMovies = payload => (
  { type: SORT_MOVIES, payload }
);

export const fetchRejected = err => (
  { type: FETCH_REJECTED, payload: { errorMessage: err } }
);

export const changeSearchBy = payload => (
  { type: SEARCH_BY, payload }
);
