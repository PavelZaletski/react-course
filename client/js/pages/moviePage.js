import React from 'react';
import { TopContainer } from '../components/topContainer';
import { connect } from 'react-redux';
import { fetchMovieById, fetchMoviesByGenres } from '../actions/movies-actions';
import MovieItems from '../components/movieItems';

export class MoviePageComponent extends React.Component {
    componentDidMount() {
        this.props.fetchMovieById(this.props.match.params.id);
    }

    render() {
        const { movie, errorMessage } = this.props;
        let genre = movie && movie.genres && movie.genres[0];

        return (
            movie ?
            <div>
                {errorMessage && <div>{errorMessage}</div>}
                <TopContainer>
                    <div className="movie-page__info">
                        <div className="movie-page__img-wrapper">
                            <img src={movie.poster_path} className="movie-page__img" />
                        </div>
                        <div className="movie-page__description">
                            <h2 className="movie-page__title">{movie.title}</h2>
                            <div className="movie-page__tagline"> {movie.tagline}</div>
                            <span className="movie-page__year">{new Date(movie.release_date).getFullYear()}</span>
                            {movie.runtime && <span className="movie-page__runtime">{`${movie.runtime} min`}</span>}
                            <div className="movie-page__overview"> {movie.overview}</div>
                        </div>
                    </div>
                </TopContainer>

                <MovieItems genre={genre}/>
            </div>
            :
            null
        );
    }
}

const mapStateToProps = (store) => ({
    movie: store.movies.selectedMovie,
    fetched: store.movies.selectedMovieFetched,
    errorMessage: store.movies.errorMessage
});

const mapDispatchToProps = (dispatch) => ({
    fetchMovieById: (id) => {
        dispatch(fetchMovieById(id))
    },

    fetchMoviesByGenres: (genres) => {
        dispatch(fetchMoviesByGenres(genres))
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(MoviePageComponent);

