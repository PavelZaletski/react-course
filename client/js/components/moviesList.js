import React from 'react';
import PropTypes from 'prop-types';
import MovieItem from './movieItem';
import MovieNotFound from './movieNotFound';
import { RadioInput } from './radioInput';

export class MoviesList extends React.Component {
    static propTypes = {
        data: PropTypes.array.isRequired
    };

    state = {
        sortValue: 'release'
    };

    sort(arr) {
        if (this.state.sortValue === 'release'){
            return arr.sort((a, b) => new Date(b.release_date) - new Date(a.release_date));
        } else {
            return arr.sort((a, b) => b.vote_average - a.vote_average);
        }
    }

    changeSorting = (value) => {
        this.setState({
            sortValue: value
        });
    }

    render() {
        const data = this.sort(this.props.data);
        const Items = data.map(item => <MovieItem item={item} key={item.id}/>);
        const { sortValue } = this.state; 

        return (
            data.length ?
            <div className="movies-list">
                <div className="movies-list__header">
                    <span>{data.length} movies found</span>
                    <div className="sortby">
                        <span>Sort by</span>
                        <RadioInput value="release" name="sort" text="Release date" onChange={this.changeSorting} checked={sortValue === 'release'}/>
                        <RadioInput value="rating" name="sort" onChange={this.changeSorting} checked={sortValue === 'rating'} />
                    </div>
                </div>
                <div className="movies-list__items">
                    {Items}
                </div>
            </div>
            :
            <MovieNotFound />
        );
    }
}
