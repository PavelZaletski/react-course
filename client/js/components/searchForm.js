import React from 'react';
import { TopContainer } from './topContainer';
import { RadioInput } from './radioInput';
import { withRouter } from 'react-router';

class SearchFormComponent extends React.Component {
    state = {
        searchBy: 'title',
        searchText: ''
    };

    onSearchTextChange = (e) => {
        this.setState({
            searchText: e.target.value
        });
    }

    onRadioButtonChange = (searchBy) => {
        this.setState({
            searchBy
        });
    }

    onSubmit = (e) => {
        e.preventDefault();
        const { searchBy, searchText } = this.state;

        this.props.history.push(`/search/${searchBy}/${searchText}`);
    }

    render() {
        const { searchBy } = this.state;
        return (
            <TopContainer>
                <form onSubmit={this.onSubmit}>
                    <div className="search-hint">Find Your movie</div>
                    <input className="search-input" type="text" name="title" required="required" placeholder="Quentin Tarantino" onChange={this.onSearchTextChange}/>

                    <div className="form-footer">
                        <div className="searchby">
                            <span className="filter-text">Search by</span>
                            <RadioInput value="title" name="search" onChange={this.onRadioButtonChange} checked={searchBy === 'title'} />
                            <RadioInput value="genres" name="search" onChange={this.onRadioButtonChange} checked={searchBy === 'genres'} />
                        </div>
                        <input className="search-button btn" type="submit" value="search" required="required" />
                    </div>
                </form>
            </TopContainer>
        );
    }
}

export const SearchForm = withRouter(SearchFormComponent);