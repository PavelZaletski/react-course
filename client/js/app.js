import React from 'react';
import ReactDOM from 'react-dom';
import { SearchForm } from './components/searchForm';
import { MoviesList } from './components/moviesList';
import { fetchMovies } from './actions/movies-actions';
import { MoviePage } from './components/moviePage';
import { ErrorBoundary } from './components/errorBoundary';
import Footer from './components/footer';

export class App extends React.Component {
    state = {
        movies: [],
        fetched: false
    };

    searchHandler = (params) => {
        fetchMovies(params).then(({data: {data}}) => {
            this.setState({
                movies: data,
                fetched: true
            });
        });
    }

    render() {
        return (
            <ErrorBoundary>
                <SearchForm onSearch={this.searchHandler}/>
                {this.state.fetched && <MoviesList data={this.state.movies} /> }
                {this.state.fetched && <MoviePage movie={this.state.movies[0]} />}
                <Footer />
            </ErrorBoundary>
        );
    }
}
