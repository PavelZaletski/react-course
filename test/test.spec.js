import renderer from 'react-test-renderer';
import Enzyme, { shallow, mount } from 'enzyme';
import React from 'react';
import Footer from '../client/js/components/footer';
import { RadioInput } from '../client/js/components/radioInput';
import MovieItem from '../client/js/components/movieItem';
import { MoviePage } from '../client/js/components/moviePage';
import { TopContainer } from '../client/js/components/topContainer';
import { SearchForm } from '../client/js/components/searchForm';
import MovieNotFound from '../client/js/components/movieNotFound';
import { ErrorBoundary } from '../client/js/components/errorBoundary';
import { MoviesList } from '../client/js/components/moviesList';
import { App } from '../client/js/app';
import Adapter from 'enzyme-adapter-react-16';
import { fetchMovies } from '../client/js/actions/movies-actions';

Enzyme.configure({ adapter: new Adapter()})

describe('Footer', () => {
    it('should match snapshot', () => {
        const component = renderer.create(<Footer />).toJSON();

        expect(component).toMatchSnapshot();
    });

    it('should change data', () => {
        const eventMock = {
            target: {}
        };

        const mockCallback = jest.fn();
        const value = 'test1';
        const name = 'test2';
        const component = shallow(<RadioInput onChange={mockCallback} name={name} value={value} />);

        component.find('input').simulate('change', eventMock);
        expect(mockCallback.mock.calls.length).toBe(1);

        const textValue = component
            .find('span')
            .text();

        expect(textValue).toEqual(value);
    });
});

describe('MovieItem', () => {
    it('should match snapshot', () => {
        const item = {
            genres: ['comedy', 'drama'],
            poster_path: 'test',
            title: 'title',
            release_date: '2015-01-01'
        };

        const component = renderer.create(<MovieItem item={item}/>).toJSON();

        expect(component).toMatchSnapshot();
    });
});

describe('MoviePage', () => {
    it('should match snapshot', () => {
        let component = renderer.create(<MoviePage />).toJSON();
        expect(component).toMatchSnapshot();
    });

    it('should match snapshot', () => {
        const item = {
            poster_path: 'test',
            title: 'title',
            release_date: '2015-01-01',
            tagline: 'some description...',
            overview: 'lorem ipsum dolor'
        };

        let component = renderer.create(<MoviePage item={item} />).toJSON();
        expect(component).toMatchSnapshot();

        item.runtime = 100;

        component = renderer.create(<MoviePage item={item} />).toJSON();
        expect(component).toMatchSnapshot();
    });
});

describe('MovieNotFound', () => {
    it('should match snapshot', () => {
        const component = renderer.create(<MovieNotFound />).toJSON();
        expect(component).toMatchSnapshot();
    });
});

describe('ErrorBoundary', () => {
    it('should match snapshot', () => {
        const component = renderer.create(<ErrorBoundary><div>movies</div></ErrorBoundary>).toJSON();
        expect(component).toMatchSnapshot();
    });
});

describe('MoviesList', () => {
    it('should match snapshot', () => {
        const component = renderer.create(<MoviesList data={[]} />).toJSON();
        expect(component).toMatchSnapshot();
    });

    it('should match snapshot', () => {
        const data = [{ id: 1, genres: [] }, { id: 2, genres: [] }]
        const component = renderer.create(<MoviesList data={data} />).toJSON();
        expect(component).toMatchSnapshot();
    });
});

describe('TopContainer', () => {
    it('should match snapshot call callback with proper param', () => {
        const eventMock = {
            preventDefault() {}
        };

        const mockCallback = jest.fn();
        let component = shallow(<SearchForm onSearch={mockCallback} />);

        component.find('form').simulate('submit', eventMock);
        expect(mockCallback.mock.calls.length).toBe(1);
        expect(mockCallback.mock.calls[0][0]).toMatchObject({ searchBy: 'title' });
        
        component = renderer.create(<SearchForm />).toJSON();

        expect(component).toMatchSnapshot();
    });

    it('should call callback with proper param', () => {
        const eventMock = {
            target: {
                value: 'genres'
            }
        };

        const mockCallback = jest.fn();
        let component = mount(<SearchForm onSearch={mockCallback} />);

        component.find("input[value='genres']").simulate('change', eventMock);

        component.find('form').simulate('submit', eventMock);
        expect(mockCallback.mock.calls.length).toBe(1);
        expect(mockCallback.mock.calls[0][0]).toMatchObject({ searchBy: 'genres'});

    });

    it('should match snapshot', () => {
        const eventMock = {
            target: {
                value: 'test'
            }
        };

        const mockCallback = jest.fn();
        let component = mount(<SearchForm onSearch={mockCallback} />);

        component.find(".search-input").simulate('change', eventMock);

        component.find('form').simulate('submit', eventMock);
        expect(mockCallback.mock.calls.length).toBe(1);
        expect(mockCallback.mock.calls[0][0]).toMatchObject({ search: 'test' });

    });
});

