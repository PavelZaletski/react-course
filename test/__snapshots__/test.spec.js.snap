// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should match snapshot 1`] = `
<div>
  movies
</div>
`;

exports[`Footer should match snapshot 1`] = `
<footer>
  Netflixroulette
</footer>
`;

exports[`MovieItem should match snapshot 1`] = `
<div
  className="movie-item"
>
  <img
    className="movie-item__img"
    src="test"
  />
  <div
    className="movie-item__description"
  >
    <h4
      className="movie-item__title"
    >
      title
    </h4>
    <span
      className="movie-item__release-date"
    >
      2015
    </span>
  </div>
  <div
    className="movie-item__genres"
  >
    <span>
      comedy
    </span>
    <span>
      drama
    </span>
  </div>
</div>
`;

exports[`MovieNotFound should match snapshot 1`] = `
<div
  className="no-found"
>
  No films found
</div>
`;

exports[`MoviePage should match snapshot 1`] = `null`;

exports[`MoviePage should match snapshot 2`] = `null`;

exports[`MoviePage should match snapshot 3`] = `null`;

exports[`MoviesList should match snapshot 1`] = `
<div
  className="no-found"
>
  No films found
</div>
`;

exports[`MoviesList should match snapshot 2`] = `
<div
  className="movies-list"
>
  <div
    className="movies-list__header"
  >
    <span>
      2
       movies found
    </span>
    <div
      className="sortby"
    >
      <span>
        Sort by
      </span>
      <label
        className="checked"
      >
        <input
          name="sort"
          onChange={[Function]}
          type="radio"
          value="release"
        />
        <span>
          Release date
        </span>
      </label>
      <label
        className=""
      >
        <input
          name="sort"
          onChange={[Function]}
          type="radio"
          value="rating"
        />
        <span>
          rating
        </span>
      </label>
    </div>
  </div>
  <div
    className="movies-list__items"
  >
    <div
      className="movie-item"
    >
      <img
        className="movie-item__img"
        src={undefined}
      />
      <div
        className="movie-item__description"
      >
        <h4
          className="movie-item__title"
        />
        <span
          className="movie-item__release-date"
        >
          NaN
        </span>
      </div>
      <div
        className="movie-item__genres"
      />
    </div>
    <div
      className="movie-item"
    >
      <img
        className="movie-item__img"
        src={undefined}
      />
      <div
        className="movie-item__description"
      >
        <h4
          className="movie-item__title"
        />
        <span
          className="movie-item__release-date"
        >
          NaN
        </span>
      </div>
      <div
        className="movie-item__genres"
      />
    </div>
  </div>
</div>
`;

exports[`TopContainer should match snapshot call callback with proper param 1`] = `
<div
  className="top-container"
>
  <div
    className="content-wrapper"
  >
    <div
      className="content"
    >
      <div
        className="title"
      >
        Netflixroulette
      </div>
      <form
        onSubmit={[Function]}
      >
        <div
          className="search-hint"
        >
          Find Your movie
        </div>
        <input
          className="search-input"
          name="title"
          onChange={[Function]}
          placeholder="Quentin Tarantino"
          required="required"
          type="text"
        />
        <div
          className="form-footer"
        >
          <div
            className="searchby"
          >
            <span
              className="filter-text"
            >
              Search by
            </span>
            <label
              className="checked"
            >
              <input
                name="search"
                onChange={[Function]}
                type="radio"
                value="title"
              />
              <span>
                title
              </span>
            </label>
            <label
              className=""
            >
              <input
                name="search"
                onChange={[Function]}
                type="radio"
                value="genres"
              />
              <span>
                genres
              </span>
            </label>
          </div>
          <input
            className="search-button btn"
            required="required"
            type="submit"
            value="search"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
